File: backend/src/core/rbac.py
"""
Role-based access control dependencies and guards.
"""
from typing import Optional
from fastapi import Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.auth import get_current_user
from src.core.db import get_db
from src.models.user import User
from src.models.user_role import RoleType
from src.services.role_service import has_role, get_user_roles


async def require_role(
    required_role: RoleType,
    school_id: Optional[int] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> User:
    """
    Dependency to require a specific role.
    
    Args:
        required_role: Role type required
        school_id: Optional school ID for scoped roles
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Current user if they have the required role
        
    Raises:
        HTTPException: 403 if user doesn't have the required role
    """
    # Check for super-admin first (can access everything)
    is_super_admin = await has_role(db, current_user.id, RoleType.SUPER_ADMIN, None)
    if is_super_admin:
        return current_user
    
    # Check for required role
    has_required_role = await has_role(db, current_user.id, required_role, school_id)
    
    if not has_required_role:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"User does not have required role: {required_role.value}",
        )
    
    return current_user


def require_super_admin(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> User:
    """Dependency to require super-admin role."""
    async def check():
        if not await has_role(db, current_user.id, RoleType.SUPER_ADMIN, None):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Super-admin access required",
            )
        return current_user
    
    return check()


def require_admin(
    school_id: Optional[int] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> User:
    """Dependency to require admin role (or super-admin)."""
    async def check():
        # Super-admin can access everything
        if await has_role(db, current_user.id, RoleType.SUPER_ADMIN, None):
            return current_user
        
        # Check admin role for specific school
        if not await has_role(db, current_user.id, RoleType.ADMIN, school_id):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Admin access required",
            )
        
        return current_user
    
    return check()


def require_instructor(
    school_id: Optional[int] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> User:
    """Dependency to require instructor role (or higher)."""
    async def check():
        # Super-admin or admin can access
        if await has_role(db, current_user.id, RoleType.SUPER_ADMIN, None):
            return current_user
        if await has_role(db, current_user.id, RoleType.ADMIN, school_id):
            return current_user
        
        # Check instructor role
        if not await has_role(db, current_user.id, RoleType.INSTRUCTOR, school_id):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Instructor access required",
            )
        
        return current_user
    
    return check()

File: src/lib/middleware/rbac.ts
import { PrismaClient, RoleType, User } from "@prisma/client";
import { APIError } from "../errors";

const prisma = new PrismaClient();

/**
 * Check if user has a specific role
 * @param userId User ID
 * @param roleType Role type to check
 * @param schoolId Optional school ID for school-specific roles
 * @returns True if user has the role
 */
export async function hasRole(
  userId: number,
  roleType: RoleType,
  schoolId?: number | null,
): Promise<boolean> {
  const role = await prisma.userRole.findFirst({
    where: {
      userId,
      roleType,
      ...(schoolId !== undefined ? { schoolId } : {}),
    },
  });

  return !!role;
}

/**
 * Get all roles for a user
 * @param userId User ID
 * @param schoolId Optional school ID to filter by
 * @returns Array of user roles
 */
export async function getUserRoles(userId: number, schoolId?: number) {
  return await prisma.userRole.findMany({
    where: {
      userId,
      ...(schoolId ? { schoolId } : {}),
    },
    include: {
      school: true,
    },
  });
}

/**
 * Require specific role for Next.js API route
 * Throws APIError if user doesn't have the role
 * @param user User object
 * @param requiredRole Required role type
 * @param schoolId Optional school ID for school-specific roles
 */
export async function requireRole(
  user: User | null,
  requiredRole: RoleType,
  schoolId?: number | null,
): Promise<void> {
  if (!user) {
    throw new APIError(401, "Not authenticated");
  }

  // Check for super-admin first (can access everything)
  const isSuperAdmin = await hasRole(user.id, RoleType.SUPER_ADMIN, null);
  if (isSuperAdmin) {
    return;
  }

  // Check for required role
  const hasRequiredRole = await hasRole(user.id, requiredRole, schoolId);

  if (!hasRequiredRole) {
    throw new APIError(403, `User does not have required role: ${requiredRole}`);
  }
}

/**
 * Require super-admin role
 * Throws APIError if user is not a super-admin
 * @param user User object
 */
export async function requireSuperAdmin(user: User | null): Promise<void> {
  if (!user) {
    throw new APIError(401, "Not authenticated");
  }

  const isSuperAdmin = await hasRole(user.id, RoleType.SUPER_ADMIN, null);
  if (!isSuperAdmin) {
    throw new APIError(403, "Super-admin access required");
  }
}

/**
 * Require admin role
 * Super-admin can also access. Throws APIError if user doesn't have required access.
 * @param user User object
 * @param schoolId Optional school ID for school-specific admin role
 */
export async function requireAdmin(user: User | null, schoolId?: number | null): Promise<void> {
  if (!user) {
    throw new APIError(401, "Not authenticated");
  }

  // Super-admin can access everything
  const isSuperAdmin = await hasRole(user.id, RoleType.SUPER_ADMIN, null);
  if (isSuperAdmin) {
    return;
  }

  // Check admin role for specific school
  const isAdmin = await hasRole(user.id, RoleType.ADMIN, schoolId);
  if (!isAdmin) {
    throw new APIError(403, "Admin access required");
  }
}

/**
 * Require instructor role
 * Super-admin and admin can also access. Throws APIError if user doesn't have required access.
 * @param user User object
 * @param schoolId Optional school ID for school-specific instructor role
 */
export async function requireInstructor(
  user: User | null,
  schoolId?: number | null,
): Promise<void> {
  if (!user) {
    throw new APIError(401, "Not authenticated");
  }

  // Super-admin or admin can access
  const isSuperAdmin = await hasRole(user.id, RoleType.SUPER_ADMIN, null);
  if (isSuperAdmin) {
    return;
  }

  const isAdmin = await hasRole(user.id, RoleType.ADMIN, schoolId);
  if (isAdmin) {
    return;
  }

  // Check instructor role
  const isInstructor = await hasRole(user.id, RoleType.INSTRUCTOR, schoolId);
  if (!isInstructor) {
    throw new APIError(403, "Instructor access required");
  }
}

/**
 * Require student role
 * Super-admin, admin, and instructor can also access. Throws APIError if user doesn't have required access.
 * @param user User object
 * @param schoolId Optional school ID for school-specific student role
 */
export async function requireStudent(user: User | null, schoolId?: number | null): Promise<void> {
  if (!user) {
    throw new APIError(401, "Not authenticated");
  }

  // Super-admin or admin can access
  const isSuperAdmin = await hasRole(user.id, RoleType.SUPER_ADMIN, null);
  if (isSuperAdmin) {
    return;
  }

  const isAdmin = await hasRole(user.id, RoleType.ADMIN, schoolId);
  if (isAdmin) {
    return;
  }

  const isInstructor = await hasRole(user.id, RoleType.INSTRUCTOR, schoolId);
  if (isInstructor) {
    return;
  }

  // Check student role
  const isStudent = await hasRole(user.id, RoleType.STUDENT, schoolId);
  if (!isStudent) {
    throw new APIError(403, "Student access required");
  }
}
